[Di Sheng] iP

dishenggg opened the pull request:
# Joe
> “Inside you there ~~are two~~ is one president~~s~~, It's Joever ~~and We're Barrack~~.” –  ([source](https://i.redd.it/the-duality-of-a-president-v0-qbi83gl0nuua1.png?s=9bf498f02f0380a9fa86b2249fe437d7a78dd489))

Joe frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~simple~~ **FAST SUPER FAST** to use

All you need to do is,

1. [view the repo here](https://github.com/dishenggg/ip/).
2. ?????
3. profit 💰.

And it is **FREE**!

## Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you're a **Java programmer**, you can use it to practice Java too. Here's the `main` method:

```java
public class Joe {
    public static void main(String[] args) {
        new Joe("joe.txt").run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
Add greet and exit to chat bot

dishenggg committed:
Level 1. Echo

dishenggg committed:
Added list to chatbot

dishenggg committed:
Added mark as done to chatbot

dishenggg committed:
Added Todos, events and deadlines

dishenggg committed:
Added Automated Testing

dishenggg committed:
Add bye to test input

dishenggg committed:
Clean up

dishenggg committed:
Added exceptions

dishenggg committed:
Updated tests

dishenggg committed:
Added delete

dishenggg committed:
Removed DukeException.java and changed formatting

dishenggg committed:
Change IOBE to DukeIOBException

dishenggg committed:
Changed commands to use enum

dishenggg committed:
Clean up imports

dishenggg committed:
Add saving, update test input.txt

dishenggg committed:
Merge branch 'branch-Level-7'

dishenggg committed:
Change deadline and event input to datetime

dishenggg committed:
Update EXPECTED.TXT and input.txt

dishenggg committed:
Merge branch 'branch-Level-8'

dishenggg committed:
Rename Duke to Joe

dishenggg committed:
Split Joe into more objects

dishenggg committed:
Merge branch 'branch-A-MoreOOP'

dishenggg committed:
Divide classes into packages

dishenggg committed:
Merge branch 'branch-A-Packages'

dishenggg committed:
Merge remote-tracking branch 'origin/add-gradle-support'

dishenggg committed:
Set up gradle for joe

dishenggg committed:
Fix regex for some commands

dishenggg committed:
Add unit tests for Parser.parse()

dishenggg committed:
Add unit tests for TaskList.toString()

dishenggg committed:
Merge branch 'branch-A-Junit'

dishenggg committed:
Configure gradle for JAR file

dishenggg committed:
Add javadocs

dishenggg committed:
Change formatting to follow coding standards

dishenggg committed:
Add find

dishenggg committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/joe/Joe.java
#	src/main/java/joe/Parser.java
#	src/main/java/joe/Storage.java
#	src/main/java/joe/TaskList.java
#	src/main/java/joe/Ui.java
#	src/main/java/joe/commands/ByeCommand.java
#	src/main/java/joe/commands/Command.java
#	src/main/java/joe/commands/DeadlineCommand.java
#	src/main/java/joe/commands/DeleteCommand.java
#	src/main/java/joe/commands/EventCommand.java
#	src/main/java/joe/commands/InvalidCommand.java
#	src/main/java/joe/commands/ListCommand.java
#	src/main/java/joe/commands/MarkCommand.java
#	src/main/java/joe/commands/TodoCommand.java
#	src/main/java/joe/commands/UnmarkCommand.java
#	src/main/java/joe/exceptions/JoeException.java
#	src/main/java/joe/exceptions/JoeIndexOutOfBoundsException.java
#	src/main/java/joe/tasks/DeadlineTask.java
#	src/main/java/joe/tasks/EventTask.java
#	src/main/java/joe/tasks/Task.java
#	src/main/java/joe/tasks/TodoTask.java

dishenggg committed:
Fix merge error

dishenggg committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/joe/Parser.java
#	src/main/java/joe/TaskList.java

dishenggg committed:
Add test cases and fix case for find

dishenggg committed:
Add javadocs for Parser.java

dishenggg committed:
Merge branch 'branch-A-JavaDoc'

raynertjx commented on src/main/java/joe/Parser.java:
```diff
@@ -0,0 +1,193 @@
+package joe;
+
+import joe.commands.ByeCommand;
+import joe.commands.Command;
+import joe.commands.DeadlineCommand;
+import joe.commands.DeleteCommand;
+import joe.commands.EventCommand;
+import joe.commands.FindCommand;
+import joe.commands.InvalidCommand;
+import joe.commands.ListCommand;
+import joe.commands.MarkCommand;
+import joe.commands.TodoCommand;
+import joe.commands.UnmarkCommand;
+import joe.exceptions.JoeException;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
```
Good job on not using wildcard imports! 😄

raynertjx commented on src/main/java/joe/Parser.java:
```diff
@@ -0,0 +1,193 @@
+        list, todo, deadline, event, mark, unmark, delete, bye, find, INVALID
+    }
+
+    private static final Pattern COMMAND_PATTERN = Pattern.compile("^(\\S+)\\s?(.*)$");
+    private static final String DATETIME_FORMAT = "d/M/yyyy HHmm";
+    private static final String FAILED_TO_PARSE_DATE_MESSAGE = "Failed to parse the date.\nPlease ensure it is a valid datetime following the format <d/M/yyyy HHmm>";
+    private static final String FAILED_TO_PARSE_INDEX_MESSAGE = "Failed to parse index";
```
Good use of all uppercase for constants.

raynertjx commented on src/main/java/joe/Parser.java:
```diff
@@ -0,0 +1,193 @@
+        CommandType type = parseType(m.group(1));
+        String args = m.group(2).trim();
+
+        switch (type) {
+        case bye:
+            if (!args.isEmpty()) {
+                return new InvalidCommand("Invalid Command! Too many arguments!");
+            }
+            return new ByeCommand();
+        case list:
+            if (!args.isEmpty()) {
+                return new InvalidCommand("Invalid Command! Too many arguments!");
+            }
+            return new ListCommand();
+        case mark:
+            return handleMark(args);
+        case unmark:
+            return handleUnmark(args);
+        case todo:
+            return handleTodo(args);
+        case deadline:
+            return handleDeadline(args);
+        case event:
+            return handleEvent(args);
+        case delete:
+            return handleDelete(args);
+        case find:
+            return handleFind(args);
+        default:
+            return handleInvalidKeyword();
+        }
+    }
```
Switch block has correct indentation!

raynertjx commented on src/main/java/joe/commands/MarkCommand.java:
```diff
@@ -0,0 +1,42 @@
+ * Represents a command to mark a task as done in the task list.
+ */
+public class MarkCommand extends Command {
+    private final int idx;
```
Perhaps we could use a more intuitive variable name here? Something like `taskIndex`.

raynertjx commented on src/main/java/joe/commands/InvalidCommand.java:
```diff
@@ -0,0 +1,33 @@
+ * Represents a command for handling invalid user inputs.
+ */
+public class InvalidCommand extends Command {
+    private final String msg;
```
Perhaps we could use a more intuitive variable name here? Something like `errorMessage`.

raynertjx reviewed (commented):
LGTM! Just some nits to fix. You've mostly kept to the coding standard which is good. 😄

dishenggg committed:
Add check style to gradle

dishenggg committed:
Merge branch 'branch-A-CheckStyle'

dishenggg committed:
Reformat code to adhere to checkstyle

dishenggg committed:
Merge branch 'branch-A-CheckStyle'

dishenggg committed:
Update build.gradle for JavaFX

dishenggg committed:
Add basic ui for joe

dishenggg committed:
Change joe to use GUI

dishenggg committed:
Merge branch 'branch-Level-10'

jianyangg commented on src/main/java/joe/Parser.java:
```diff
@@ -0,0 +1,210 @@
+package joe;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import joe.commands.ByeCommand;
+import joe.commands.Command;
+import joe.commands.DeadlineCommand;
+import joe.commands.DeleteCommand;
+import joe.commands.EventCommand;
+import joe.commands.FindCommand;
+import joe.commands.InvalidCommand;
+import joe.commands.ListCommand;
+import joe.commands.MarkCommand;
+import joe.commands.TodoCommand;
+import joe.commands.UnmarkCommand;
+import joe.exceptions.JoeException;
```
Consider arranging your imports in lexicographical order, which could be set-up as part of your IDE's check style  process.

jianyangg commented on src/main/java/joe/ui/DialogBox.java:
```diff
@@ -0,0 +1,95 @@
+package joe.ui;
+
+import java.io.IOException;
+import java.util.Collections;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import javafx.fxml.FXML;
+import javafx.fxml.FXMLLoader;
+import javafx.geometry.Pos;
+import javafx.scene.Node;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+import javafx.scene.shape.Rectangle;
+import javafx.scene.text.Text;
+import javafx.scene.text.TextAlignment;
+
```
See above

jianyangg commented on src/main/java/joe/commands/Command.java:
```diff
@@ -0,0 +1,31 @@
+     * @param storage The storage for saving and loading tasks.
+     * @throws JoeIndexOutOfBoundsException If there is an invalid index used in the command.
+     */
+    public abstract String execute(TaskList tasks, Storage storage)
```
Consider using a better naming for this function. `execute` sounds slightly vague.

jianyangg reviewed (commented):
LGTM! Code is generally written quite well, just nitpicking a little.

dishenggg committed:
Fix image filename

dishenggg committed:
Fix error message on empty input

dishenggg committed:
Merge branch 'branch-Level-10'

dishenggg committed:
Set up CI

dishenggg committed:
Add assertion to check for TaskList size after command

There are several commands whose main purpose is to add or remove items
from a TaskList.

Assertions are added to validate that the size of the TaskList changed
accordingly before saving to file.

dishenggg committed:
Change execute to throw JoeException

Previously execute only throws JoeIndexOutOfBoundsException which does
not account for all possible exceptions that can happen during the
execution of a command.

Changing the abstract execute method in Command.java to throw
JoeException instead will allow it to handle more exceptions.

dishenggg committed:
Add assertions

dishenggg committed:
Refactor code to improve code quality

dishenggg committed:
Merge pull request #1 from dishenggg/branch-A-Assertions

Add assertions to multiple areas

dishenggg committed:
Merge branch 'master' into branch-A-CodeQuality

dishenggg committed:
Merge pull request #2 from dishenggg/branch-A-CodeQuality

Refactor code to improve code quality

dishenggg committed:
Merge branch 'master' into branch-A-CI

dishenggg committed:
Merge pull request #3 from dishenggg/branch-A-CI

Add CI

dishenggg committed:
Rename parse to parseUserInput for clarity

dishenggg committed:
Fix bug in Storage.java

dishenggg committed:
Add findmatch and findall

dishenggg committed:
Merge pull request #4 from dishenggg/branch-C-BetterSearch

Add more search options

dishenggg committed:
Edit code style

dishenggg committed:
Fix Ui for bye command

dishenggg committed:
Update Joe User Guide

dishenggg committed:
Update TaskListTest.java to match format change

dishenggg committed:
Update TaskListTest.java to be independent of Task

dishenggg committed:
Add tests for Commands

dishenggg committed:
Update user guide formatting

dishenggg committed:
Update user guide formatting
